### PR TITLE
fix: allow to collect DS information per operation and ensure corrected errors are passed

### PR DESCRIPTION
**Pull request by StarpTech**

### CODE REVIEW

File: v2/pkg/engine/postprocess/postprocess.go [hunk 1]
```diff
@@ -161,3 +177,17 @@ func (p *Processor) createFetchTree(res *resolve.GraphQLResponse) {
 		ChildNodes: children,
 	}
 }
+
+// collectDataSourceInfos returns the list of involved data sources of the operation
+func collectDataSourceInfos(node *resolve.FetchTreeNode) (list []resolve.DataSourceInfo) {
+
```
Comment (df-wg) [inline v2/pkg/engine/postprocess/postprocess.go]: nit: remove extra line

File: v2/pkg/engine/postprocess/postprocess_test.go [hunk 2]
```diff
@@ -334,7 +335,281 @@ func TestProcess_ExtractFetches(t *testing.T) {
...
+						Fields: []*resolve.Field{
+							{
+								Name: []byte("field1"),
+								Value: &resolve.String{
+									Path: []string{"field1"},
+								},
+							},
+						},
+						Fetches: []resolve.Fetch{
+
```
Comment (df-wg) [inline v2/pkg/engine/postprocess/postprocess_test.go]: nit: extra line

**Review by df-wg — APPROVED** [review]
LGTM

How will this be used in the router?

File: v2/pkg/engine/postprocess/postprocess.go [hunk 3]
```diff
@@ -161,3 +177,17 @@ func (p *Processor) createFetchTree(res *resolve.GraphQLResponse) {
...
+	if node.Item != nil && node.Item.Fetch != nil {
+		list = append(list, node.Item.Fetch.DataSourceInfo())
+	}
+
+	for _, childNode := range node.ChildNodes {
+		list = append(list, collectDataSourceInfos(childNode)...)
+	}
+
+	return unique.SliceElements(list)
```
Comment (devsergiy) [inline v2/pkg/engine/postprocess/postprocess.go]: could we use https://pkg.go.dev/slices#Compact?

Comment (StarpTech) [reply v2/pkg/engine/postprocess/postprocess.go]: Nice, never heard about it.

**Comment by StarpTech** [comment]
> How will this be used in the router?

It makes the involved datasource infos available after the post processing step. This info is used to enrich metrics.

File: v2/pkg/engine/postprocess/postprocess.go [hunk 4]
```diff
@@ -161,3 +177,16 @@ func (p *Processor) createFetchTree(res *resolve.GraphQLResponse) {
 		ChildNodes: children,
 	}
 }
+
+// collectDataSourceInfos returns the list of involved data sources of the operation
+func collectDataSourceInfos(node *resolve.FetchTreeNode) (list []resolve.DataSourceInfo) {
```
Comment (devsergiy) [inline v2/pkg/engine/postprocess/postprocess.go]: btw, you could do this in the createFetchTree

there is a step when you got all fetches slice
ao you could just iterate and collect datasource in one pass without the need to traverse the tree

Comment (StarpTech) [reply v2/pkg/engine/postprocess/postprocess.go]: Great suggestion. I'll change it.

**Review by devsergiy — APPROVED** [review]

**Pull request merged**